subxt/src/runtime_api/runtime_payload.rs

```diff
@@ -0,0 +1,185 @@
+// Copyright 2019-2023 Parity Technologies (UK) Ltd.
+// This file is dual-licensed as Apache-2.0 or GPL-3.0.
+// see LICENSE for license details.
+
+use codec::Decode;
+use core::marker::PhantomData;
+use scale_encode::EncodeAsFields;
+use scale_value::Composite;
+use std::borrow::Cow;
+
+use crate::dynamic::DecodedValueThunk;
+use crate::{metadata::DecodeWithMetadata, Error, Metadata};
+
+/// This represents a runtime API payload that can call into the runtime of node.
+///
+/// # Components
+///
+/// - associated return type
+///
+/// Resulting bytes of the call are interpreted into this type.
+///
+/// - runtime function name
+///
+/// The function name of the runtime API call. This is obtained by concatenating
+/// the runtime trait name with the trait's method.
+///
+/// For example, the substrate runtime trait [Metadata](https://github.com/paritytech/substrate/blob/cb954820a8d8d765ce75021e244223a3b4d5722d/primitives/api/src/lib.rs#L745)
+/// contains the `metadata_at_version` function. The corresponding runtime function
+/// is `Metadata_metadata_at_version`.
+///
+/// - encoded arguments
+///
+/// Each argument of the runtime function must be scale-encoded.
+pub trait RuntimeApiPayload {
+    /// The target type of the return value.
+    // Note: `DecodeWithMetadata` bound implied by the `encode_as_fields_to`.
+    type ReturnType: DecodeWithMetadata;
```

Comment:
Ideally, this associated type would removed from the trait and the impls has to specify this bound such as.

```rust
impl<D: DecodeWithMetadata> RuntimeApiPayload for DynamicRuntimeApiPayload<D> {
    fn fn_name(&self) -> &str {
        &self.fn_name
    }

    fn encode_args_to(&self, metadata: &Metadata, out: &mut Vec<u8>) -> Result<(), Error> {
        let fn_metadata = metadata.runtime_fn(&self.fn_name)?;

        self.args_data
            .encode_as_fields_to(fn_metadata.fields(), metadata.types(), out)?;

        Ok(())
    }
}
```

Doesn't matter that much to me as this is an internal trait used by subxt and shouldn't be needed to impl by "subxt users"

Reply:
Up to you but would be good to document `ReturnType` such as (I might be wrong here but just as example):

> The `ReturnType` is not needed by the trait itself but dynamic implementations might use `scale_value::encode_as_field_to` which requires this `DecodeWithMetadata trait bound`

Reply:
The `ReturnType` associated type is needed so that the Subxt API call fn can look like:

```rust
fn call<P: RuntimeApiPayload>(&self, payload: P) -> Result<P::ReturnType,..>
```

ie so that the subxt API knows what return type to decode the result into and hand back.

Statically, we know what the type should be (and codegen it). But We can also return some dynamic type by setting `ReturnType` to `DecodedValueThunk` or something :)